File: apps/mobile/modules/kssenger-signal/ios-draft/SignalDeviceProtocol.swift
import Foundation
import LibSignalClient

/// iOS port of the Android `SignalDeviceProtocol`.
///
/// Same public surface: provision / processRemoteBundle / hasSession / encrypt /
/// decrypt, same wire JSON shapes, same base64 (no-wrap) encoding. No custom
/// cryptography — only the official `LibSignalClient` primitives.
///
/// NOTE: written without a macOS/Xcode toolchain. Free-function names
/// (`processPreKeyBundle`, `signalEncrypt`, `signalDecrypt*`) and record
/// initializers follow the 0.x Swift API; verify against the pinned version.
final class SignalDeviceProtocolIOS {
  enum ProtocolError: Error {
    case invalidDeviceUuid
    case invalidSignalDeviceId
    case invalidPreKeyCount
    case plaintextTooLarge
    case ciphertextTooLarge
    case unsupportedMessageType
  }

  private let normalizedDeviceUuid: String
  private let blobs: KeychainBlobStore
  private let identity: KssengerIdentityStore
  private let sessions: KssengerSessionStore
  private let preKeys: KssengerPreKeyStore
  private let signedPreKeys: KssengerSignedPreKeyStore
  private let kyberPreKeys: KssengerKyberPreKeyStore
  private let ctx = NullContext()  // K-ssenger stores ignore StoreContext; a shared no-op is fine.

  init(deviceUuid: String) throws {
    guard let uuid = UUID(uuidString: deviceUuid) else { throw ProtocolError.invalidDeviceUuid }
    self.normalizedDeviceUuid = uuid.uuidString.lowercased()
    self.blobs = KeychainBlobStore(namespace: "signal-device-\(normalizedDeviceUuid)")
    self.identity = KssengerIdentityStore(blobs)
    self.sessions = KssengerSessionStore(blobs)
    self.preKeys = KssengerPreKeyStore(blobs)
    self.signedPreKeys = KssengerSignedPreKeyStore(blobs)
    self.kyberPreKeys = KssengerKyberPreKeyStore(blobs)
  }

  // MARK: helpers

  private func b64(_ bytes: [UInt8]) -> String { Data(bytes).base64EncodedString() }
  private func b64(_ data: Data) -> String { data.base64EncodedString() }
  private func decode(_ s: String) throws -> [UInt8] {
    guard let d = Data(base64Encoded: s) else { throw ProtocolError.unsupportedMessageType }
    return [UInt8](d)
  }

  private func address(_ userId: String, _ deviceId: Int) throws -> ProtocolAddress {
    guard let uuid = UUID(uuidString: userId) else { throw ProtocolError.invalidDeviceUuid }
    guard deviceId >= 1, deviceId <= 127 else { throw ProtocolError.invalidSignalDeviceId }
    return try ProtocolAddress(name: uuid.uuidString.lowercased(), deviceId: UInt32(deviceId))
  }

  private func nextKeyId(_ excluded: Set<UInt32>) -> UInt32 {
    for _ in 0..<1000 {
      let candidate = UInt32.random(in: 1...(UInt32.max - 1))
      if !excluded.contains(candidate) { return candidate }
    }
    return UInt32.random(in: 1...(UInt32.max - 1))
  }

  // MARK: provision

  func provision(oneTimeCount: Int = 20) throws -> [String: Any] {
    guard oneTimeCount >= 10, oneTimeCount <= 100 else { throw ProtocolError.invalidPreKeyCount }
    let identityPair = try identity.identityKeyPair(context: ctx)
    let registrationId = try identity.localRegistrationId(context: ctx)
    let now = UInt64(Date().timeIntervalSince1970 * 1000)

    var used = Set<UInt32>()

    let signedId = nextKeyId(used); used.insert(signedId)
    let signedPriv = PrivateKey.generate()
    let signedPub = signedPriv.publicKey
    let signedSig = identityPair.privateKey.generateSignature(message: signedPub.serialize())
    let signedRecord = try SignedPreKeyRecord(id: signedId, timestamp: now, privateKey: signedPriv, signature: signedSig)
    try signedPreKeys.storeSignedPreKey(signedRecord, id: signedId, context: ctx)

    let lastResortId = nextKeyId(used); used.insert(lastResortId)
    let lastResortPair = KEMKeyPair.generate()
    let lastResortSig = identityPair.privateKey.generateSignature(message: lastResortPair.publicKey.serialize())
    let lastResortRecord = try KyberPreKeyRecord(id: lastResortId, timestamp: now, keyPair: lastResortPair, signature: lastResortSig)
    try kyberPreKeys.storeLastResortKyberPreKey(lastResortRecord, id: lastResortId, context: ctx)

    var ecPublic: [[String: Any]] = []
    var pqPublic: [[String: Any]] = []
    for _ in 0..<oneTimeCount {
      let ecId = nextKeyId(used); used.insert(ecId)
      let ecPriv = PrivateKey.generate()
      let ecRecord = try PreKeyRecord(id: ecId, privateKey: ecPriv)
      try preKeys.storePreKey(ecRecord, id: ecId, context: ctx)
      ecPublic.append(["keyId": ecId, "publicKey": b64(ecPriv.publicKey.serialize())])

      let pqId = nextKeyId(used); used.insert(pqId)
      let pqPair = KEMKeyPair.generate()
      let pqSig = identityPair.privateKey.generateSignature(message: pqPair.publicKey.serialize())
      let pqRecord = try KyberPreKeyRecord(id: pqId, timestamp: now, keyPair: pqPair, signature: pqSig)
      try kyberPreKeys.storeKyberPreKey(pqRecord, id: pqId, context: ctx)
      pqPublic.append(["keyId": pqId, "publicKey": b64(pqPair.publicKey.serialize()), "signature": b64(pqSig)])
    }

    let bundleVersion = max(1, Int(Date().timeIntervalSince1970) & 0x7fffffff)
    return [
      "bundleVersion": bundleVersion,
      "registrationId": registrationId,
      "identityKey": b64(identityPair.identityKey.serialize()),
      "signedPreKeyId": signedId,
      "signedPreKeyPublic": b64(signedPub.serialize()),
      "signedPreKeySignature": b64(signedSig),
      "pqLastResortPreKeyId": lastResortId,
      "pqLastResortPreKeyPublic": b64(lastResortPair.publicKey.serialize()),
      "pqLastResortPreKeySignature": b64(lastResortSig),
      "oneTimePreKeys": ecPublic,
      "pqOneTimePreKeys": pqPublic,
    ]
  }

  // MARK: processRemoteBundle

  func processRemoteBundle(
    localUserId: String, localSignalDeviceId: Int,
    remoteUserId: String, remoteSignalDeviceId: Int,
    registrationId: Int,
    identityKeyB64: String,
    signedPreKeyId: Int, signedPreKeyPublicB64: String, signedPreKeySignatureB64: String,
    oneTimePreKeyId: Int?, oneTimePreKeyPublicB64: String?,
    pqPreKeyId: Int, pqPreKeyPublicB64: String, pqPreKeySignatureB64: String
  ) throws {
    let localAddress = try address(localUserId, localSignalDeviceId)
    let remoteAddress = try address(remoteUserId, remoteSignalDeviceId)

    let identityKey = try IdentityKey(bytes: try decode(identityKeyB64))
    let signedPub = try PublicKey(try decode(signedPreKeyPublicB64))
    let signedSig = try decode(signedPreKeySignatureB64)
    let kyberPub = try KEMPublicKey(try decode(pqPreKeyPublicB64))
    let kyberSig = try decode(pqPreKeySignatureB64)

    let bundle: PreKeyBundle
    if let otId = oneTimePreKeyId, let otPubB64 = oneTimePreKeyPublicB64 {
      let otPub = try PublicKey(try decode(otPubB64))
      bundle = try PreKeyBundle(
        registrationId: UInt32(registrationId),
        deviceId: UInt32(remoteSignalDeviceId),
        prekeyId: UInt32(otId),
        prekey: otPub,
        signedPrekeyId: UInt32(signedPreKeyId),
        signedPrekey: signedPub,
        signedPrekeySignature: signedSig,
        identity: identityKey,
        kyberPrekeyId: UInt32(pqPreKeyId),
        kyberPrekey: kyberPub,
        kyberPrekeySignature: kyberSig
      )
    } else {
      bundle = try PreKeyBundle(
        registrationId: UInt32(registrationId),
        deviceId: UInt32(remoteSignalDeviceId),
        signedPrekeyId: UInt32(signedPreKeyId),
        signedPrekey: signedPub,
        signedPrekeySignature: signedSig,
        identity: identityKey,
        kyberPrekeyId: UInt32(pqPreKeyId),
        kyberPrekey: kyberPub,
        kyberPrekeySignature: kyberSig
      )
    }

    _ = localAddress
    try processPreKeyBundle(
      bundle,
      for: remoteAddress,
      sessionStore: sessions,
      identityStore: identity,
      context: ctx
    )
  }

  // MARK: session state

  func hasSession(remoteUserId: String, remoteSignalDeviceId: Int) throws -> Bool {
    let addr = try address(remoteUserId, remoteSignalDeviceId)
    return sessions.containsSession(for: addr)
  }

  // MARK: encrypt / decrypt

  func encrypt(
    localUserId: String, localSignalDeviceId: Int,
    remoteUserId: String, remoteSignalDeviceId: Int,
    plaintext: String
  ) throws -> [String: Any] {
    let data = Array(plaintext.utf8)
    guard data.count <= 64 * 1024 else { throw ProtocolError.plaintextTooLarge }
    let remoteAddress = try address(remoteUserId, remoteSignalDeviceId)
    _ = try address(localUserId, localSignalDeviceId)

    let message = try signalEncrypt(
      message: data,
      for: remoteAddress,
      sessionStore: sessions,
      identityStore: identity,
      context: ctx
    )
    let kind: String
    switch message.messageType {
    case .preKey: kind = "prekey"
    case .whisper: kind = "signal"
    default: throw ProtocolError.unsupportedMessageType
    }
    return [
      "kind": kind,
      "ciphertext": b64(message.serialize()),
      "algorithm": "signal-libsignal-v3-pqxdh",
    ]
  }

  func decrypt(
    localUserId: String, localSignalDeviceId: Int,
    remoteUserId: String, remoteSignalDeviceId: Int,
    kind: String, ciphertextB64: String
  ) throws -> String {
    let bytes = try decode(ciphertextB64)
    guard bytes.count <= 2_000_000 else { throw ProtocolError.ciphertextTooLarge }
    let remoteAddress = try address(remoteUserId, remoteSignalDeviceId)
    _ = try address(localUserId, localSignalDeviceId)

    let clear: [UInt8]
    switch kind {
    case "prekey":
      let msg = try PreKeySignalMessage(bytes: bytes)
      clear = try signalDecryptPreKey(
        message: msg,
        from: remoteAddress,
        sessionStore: sessions,
        identityStore: identity,
        preKeyStore: preKeys,
        signedPreKeyStore: signedPreKeys,
        kyberPreKeyStore: kyberPreKeys,
        context: ctx
      )
    case "signal":
      let msg = try SignalMessage(bytes: bytes)
      clear = try signalDecrypt(
        message: msg,
        from: remoteAddress,
        sessionStore: sessions,
        identityStore: identity,
        context: ctx
      )
    default:
      throw ProtocolError.unsupportedMessageType
    }
    return String(decoding: clear, as: UTF8.self)
  }

  func clearState() throws {
    try blobs.clearAndDestroyKey()
  }

  // Probes used by getStatus().
  static func libsignalLoaded() -> Bool {
    _ = IdentityKeyPair.generate()
    return true
  }
}

File: apps/mobile/modules/kssenger-signal/android/src/main/java/com/kahdigital/kssenger/signal/SignalDeviceProtocol.kt
package com.kahdigital.kssenger.signal

import android.content.Context
import android.util.Base64
import java.security.SecureRandom
import java.util.UUID
import org.signal.libsignal.protocol.IdentityKey
import org.signal.libsignal.protocol.SessionBuilder
import org.signal.libsignal.protocol.SessionCipher
import org.signal.libsignal.protocol.SignalProtocolAddress
import org.signal.libsignal.protocol.ecc.ECKeyPair
import org.signal.libsignal.protocol.ecc.ECPublicKey
import org.signal.libsignal.protocol.kem.KEMKeyPair
import org.signal.libsignal.protocol.kem.KEMKeyType
import org.signal.libsignal.protocol.kem.KEMPublicKey
import org.signal.libsignal.protocol.message.PreKeySignalMessage
import org.signal.libsignal.protocol.message.SignalMessage
import org.signal.libsignal.protocol.state.KyberPreKeyRecord
import org.signal.libsignal.protocol.state.PreKeyBundle
import org.signal.libsignal.protocol.state.PreKeyRecord
import org.signal.libsignal.protocol.state.SignedPreKeyRecord

internal class SignalDeviceProtocol(
  context: Context,
  deviceUuid: String,
) {
  private val normalizedDeviceUuid = UUID.fromString(deviceUuid).toString()
  private val blobs = KeystoreBlobStore(context, "signal-device-$normalizedDeviceUuid")
  private val identity = SignalIdentityKeyStore(blobs)
  private val sessions = SignalSessionStore(blobs)
  private val preKeys = SignalPreKeyStore(blobs)
  private val signedPreKeys = SignalSignedPreKeyStore(blobs)
  private val kyberPreKeys = SignalKyberPreKeyStore(blobs)
  private val random = SecureRandom()

  fun provision(oneTimeCount: Int = 20): Map<String, Any> {
    require(oneTimeCount in 10..100) { "INVALID_PREKEY_COUNT" }
    val identityPair = identity.identityKeyPair
    val signedId = nextKeyId()
    val signedPair = ECKeyPair.generate()
    val signedSignature = identityPair.privateKey.calculateSignature(signedPair.publicKey.serialize())
    signedPreKeys.storeSignedPreKey(
      signedId,
      SignedPreKeyRecord(signedId, System.currentTimeMillis(), signedPair, signedSignature),
    )

    val lastResortId = nextKeyId(setOf(signedId))
    val lastResortPair = KEMKeyPair.generate(KEMKeyType.KYBER_1024)
    val lastResortSignature = identityPair.privateKey.calculateSignature(lastResortPair.publicKey.serialize())
    kyberPreKeys.storeLastResortKyberPreKey(
      lastResortId,
      KyberPreKeyRecord(lastResortId, System.currentTimeMillis(), lastResortPair, lastResortSignature),
    )

    val usedIds = mutableSetOf(signedId, lastResortId)
    val ecPublic = mutableListOf<Map<String, Any>>()
    val pqPublic = mutableListOf<Map<String, Any>>()
    repeat(oneTimeCount) {
      val ecId = nextKeyId(usedIds).also(usedIds::add)
      val ecPair = ECKeyPair.generate()
      preKeys.storePreKey(ecId, PreKeyRecord(ecId, ecPair))
      ecPublic += mapOf("keyId" to ecId, "publicKey" to b64(ecPair.publicKey.serialize()))

      val pqId = nextKeyId(usedIds).also(usedIds::add)
      val pqPair = KEMKeyPair.generate(KEMKeyType.KYBER_1024)
      val pqSignature = identityPair.privateKey.calculateSignature(pqPair.publicKey.serialize())
      kyberPreKeys.storeKyberPreKey(
        pqId,
        KyberPreKeyRecord(pqId, System.currentTimeMillis(), pqPair, pqSignature),
      )
      pqPublic += mapOf(
        "keyId" to pqId,
        "publicKey" to b64(pqPair.publicKey.serialize()),
        "signature" to b64(pqSignature),
      )
    }

    val bundleVersion = ((System.currentTimeMillis() / 1000L) and 0x7fffffffL).toInt().coerceAtLeast(1)
    return mapOf(
      "bundleVersion" to bundleVersion,
      "registrationId" to identity.localRegistrationId,
      "identityKey" to b64(identityPair.publicKey.serialize()),
      "signedPreKeyId" to signedId,
      "signedPreKeyPublic" to b64(signedPair.publicKey.serialize()),
      "signedPreKeySignature" to b64(signedSignature),
      "pqLastResortPreKeyId" to lastResortId,
      "pqLastResortPreKeyPublic" to b64(lastResortPair.publicKey.serialize()),
      "pqLastResortPreKeySignature" to b64(lastResortSignature),
      "oneTimePreKeys" to ecPublic,
      "pqOneTimePreKeys" to pqPublic,
    )
  }

  fun processRemoteBundle(
    localUserId: String,
    localSignalDeviceId: Int,
    remoteUserId: String,
    remoteSignalDeviceId: Int,
    registrationId: Int,
    identityKeyB64: String,
    signedPreKeyId: Int,
    signedPreKeyPublicB64: String,
    signedPreKeySignatureB64: String,
    oneTimePreKeyId: Int?,
    oneTimePreKeyPublicB64: String?,
    pqPreKeyId: Int,
    pqPreKeyPublicB64: String,
    pqPreKeySignatureB64: String,
  ) {
    val localAddress = address(localUserId, localSignalDeviceId)
    val remoteAddress = address(remoteUserId, remoteSignalDeviceId)
    val preKeyBytes = oneTimePreKeyPublicB64?.let(::decode)
    val bundle = PreKeyBundle(
      registrationId,
      remoteSignalDeviceId,
      oneTimePreKeyId ?: PreKeyBundle.NULL_PRE_KEY_ID,
      preKeyBytes?.let(::ECPublicKey),
      signedPreKeyId,
      ECPublicKey(decode(signedPreKeyPublicB64)),
      decode(signedPreKeySignatureB64),
      IdentityKey(decode(identityKeyB64)),
      pqPreKeyId,
      KEMPublicKey(decode(pqPreKeyPublicB64)),
      decode(pqPreKeySignatureB64),
    )
    SessionBuilder(sessions, preKeys, signedPreKeys, identity, remoteAddress, localAddress).process(bundle)
  }

  fun hasSession(remoteUserId: String, remoteSignalDeviceId: Int): Boolean =
    sessions.containsSession(address(remoteUserId, remoteSignalDeviceId))

  fun encrypt(
    localUserId: String,
    localSignalDeviceId: Int,
    remoteUserId: String,
    remoteSignalDeviceId: Int,
    plaintext: String,
  ): Map<String, Any> {
    require(plaintext.toByteArray(Charsets.UTF_8).size <= 64 * 1024) { "PLAINTEXT_TOO_LARGE" }
    val localAddress = address(localUserId, localSignalDeviceId)
    val remoteAddress = address(remoteUserId, remoteSignalDeviceId)
    val cipher = SessionCipher(
      sessions, preKeys, signedPreKeys, kyberPreKeys, identity, localAddress, remoteAddress,
    )
    val encrypted = cipher.encrypt(plaintext.toByteArray(Charsets.UTF_8))
    val kind = when (encrypted) {
      is PreKeySignalMessage -> "prekey"
      is SignalMessage -> "signal"
      else -> throw IllegalStateException("UNSUPPORTED_LIBSIGNAL_MESSAGE_TYPE")
    }
    return mapOf(
      "kind" to kind,
      "ciphertext" to b64(encrypted.serialize()),
      "algorithm" to "signal-libsignal-v3-pqxdh",
    )
  }

  fun decrypt(
    localUserId: String,
    localSignalDeviceId: Int,
    remoteUserId: String,
    remoteSignalDeviceId: Int,
    kind: String,
    ciphertextB64: String,
  ): String {
    val localAddress = address(localUserId, localSignalDeviceId)
    val remoteAddress = address(remoteUserId, remoteSignalDeviceId)
    val cipher = SessionCipher(
      sessions, preKeys, signedPreKeys, kyberPreKeys, identity, localAddress, remoteAddress,
    )
    val bytes = decode(ciphertextB64)
    require(bytes.size <= 2_000_000) { "CIPHERTEXT_TOO_LARGE" }
    val clear = when (kind) {
      "prekey" -> cipher.decrypt(PreKeySignalMessage(bytes))
      "signal" -> cipher.decrypt(SignalMessage(bytes))
      else -> throw IllegalArgumentException("UNSUPPORTED_LIBSIGNAL_MESSAGE_TYPE")
    }
    return clear.toString(Charsets.UTF_8)
  }

  private fun address(userId: String, deviceId: Int): SignalProtocolAddress {
    val normalizedUser = UUID.fromString(userId).toString()
    require(deviceId in 1..127) { "INVALID_SIGNAL_DEVICE_ID" }
    return SignalProtocolAddress(normalizedUser, deviceId)
  }

  private fun nextKeyId(excluded: Set<Int> = emptySet()): Int {
    repeat(1000) {
      val candidate = random.nextInt(Int.MAX_VALUE - 1) + 1
      if (!excluded.contains(candidate)) return candidate
    }
    throw IllegalStateException("PREKEY_ID_EXHAUSTED")
  }

  private fun b64(value: ByteArray): String = Base64.encodeToString(value, Base64.NO_WRAP)
  private fun decode(value: String): ByteArray = Base64.decode(value, Base64.NO_WRAP)
}
